src/vm/peimage.cpp

```diff
@@ -1104,7 +1104,10 @@ void PEImage::Load()
 
     SimpleWriteLockHolder lock(m_pLayoutLock);
 
-    _ASSERTE(m_pLayouts[IMAGE_LOADED] == NULL);
+    if (m_pLayouts[IMAGE_LOADED] != NULL)
+    {
+        return;
+    }
```

Comment:
Well, the discussion about the memory models are always fun. What can happen in theory vs. what happens in practice.

> In case, compiler can prove this, the value can be cached, which can break semantics of this code

There is no way for the compiler to prove this. Also, the lock has barriers in it - even if the compiler was somehow able to reason about the whole lock code, the barriers inside the lock have to prevent it.

If you are worried about this, you can fetch the value outside the lock using `VolatileLoad` that should prevent any dangerous optimizations that may happen here in theory.